src/tir/transforms/local_pad.cc

```diff
@@ -0,0 +1,302 @@
+      }
+    }
+    bool NoAccesses() const {
+      return !(bit_vector_[StorageType::kGlobal] || bit_vector_[StorageType::kShared] ||
+               bit_vector_[StorageType::kLocal] || bit_vector_[StorageType::kOthers]);
+    }
+    bool OnlyGlobalAccesses() const {
+      return !(bit_vector_[StorageType::kShared] || bit_vector_[StorageType::kLocal] ||
+               bit_vector_[StorageType::kOthers]) &&
+             bit_vector_[StorageType::kGlobal];
+    }
+    bool OnlyLocalOrSharedAccesses() const {
+      return !(bit_vector_[StorageType::kGlobal] || bit_vector_[StorageType::kOthers]) &&
+             (bit_vector_[StorageType::kShared] || bit_vector_[StorageType::kLocal]);
+    }
+
+   private:
+    std::array<bool, StorageType::kOthers + 1> bit_vector_ = {false};
+  };
+  AccessMarker read_marker_, write_marker_;
+  std::pair<AccessMarker, AccessMarker> Analyze_(const Stmt& stmt) {
+    VisitStmt(stmt);
+    return std::make_pair(read_marker_, write_marker_);
+  }
+
+  friend class LocalPadder;
+};
+
+/*!
+ * \brief Verify that all local variables are initialized to the same constant expression.
+ */
+class InitChecker : public StmtVisitor {
+ private:
+  void VisitStmt_(const BufferStoreNode* op) final {
+    // Read the check the RHS values, make sure that they are the same constant for all the
+    // initialization statements.
+    CheckInitValue_<IntImmNode>(op->value);
+    CheckInitValue_<FloatImmNode>(op->value);
+    return StmtVisitor::VisitStmt_(op);
+  }
+  template <typename ImmNodeType>
+  void CheckInitValue_(const PrimExpr& rhs) {
+    if (const ImmNodeType* const rhs_val = rhs.as<ImmNodeType>()) {
+      if (init_constexpr_.defined()) {
+        if (const ImmNodeType* const init_val = init_constexpr_.as<ImmNodeType>()) {
+          if (rhs_val->value != init_val->value) {
+            init_with_single_constexpr_ = false;
+          }
+        } else {
+          init_with_single_constexpr_ = false;
+        }
+      } else {
+        init_with_single_constexpr_ = true;
+        init_constexpr_ = rhs;
+      }
+    }
+  }
+  void operator()(const Stmt& stmt) {
+    StmtVisitor::operator()(stmt);
+    if (!init_with_single_constexpr_) {
+      init_constexpr_ = PrimExpr();
+    }
+  }
+
+  bool init_with_single_constexpr_ = false;
+  PrimExpr init_constexpr_;
+
+  friend class LocalPadder;
+};
+
+/*!
+ * \brief Split a predicate into inlinable and non-inlinable component.
+ *
+ *        We refer to "inlinable predicate" as
+ *
+ *            if (predicate) A = ...;
+ *            ↓
+ *            A = predicate ? ... : init_constexpr;
+ *
+ *        Note that not all predicates can be inlined. For example, if a predicate is there to guard
+ *        against out-of-boundary accesses to local/shared variables, then it cannot be inlined.
+ */
+class PredicateInliner : public StmtExprVisitor {
+ private:
+  explicit PredicateInliner(const Stmt& body_stmt) : body_stmt_(body_stmt) {}
```

Comment:
Quick question: do we assume that `body_stmt` is a BufferStore?

Reply:
Yes. We assume that the body statements where predicates are to be inlined is a `BufferStore` node.